Add PDU Node relations using OBM services data

keedya opened the pull request:
- Adds pdu relation to node only if this particular PDU is added to the obm services list

keedya commented:
@RackHD/corecommitters @dalebremner @BillyAbildgaard @srinia6 @brianparry @zyoung51

yyscamper commented on lib/jobs/pdu-node-relations.js:
```diff
@@ -0,0 +1,142 @@
+        PduRelationsJob.super_.call(this, logger, options, context, taskId);
+
+        this.nodeId = context.target || options.nodeId;
+        assert.isMongoId(this.nodeId);
```
Don't do this, The RackHD will support multi backends database,  then the nodeId will be set as uuid not MongoId.

yyscamper commented on lib/jobs/pdu-node-relations.js:
```diff
@@ -0,0 +1,142 @@
+     */
+    PduRelationsJob.prototype._run = function run() {
+        var self = this;
+        return waterline.nodes.findByIdentifier(self.nodeId)
```
Change to `needByIdentifier`. `needByIdentifier` will throw error if the node doesn't exist, however `findByIdentifier` will return an empty.

keedya replied:
@yyscamper they both work do you agree?

anhou replied:
there is a check for returned value in needByIdentifier, it's 'node' here, if don't use needByIdentifier , need a check before 'node.type' != 'pdu',  otherwise, if there's no 'node' found by 'nodeId'  there will be error like 'TypeError: Cannot read property 'type' of undefined'

yyscamper replied:
@keedya : Either is OK. 'findByIdentifer' + Output Check = 'needByIdentifier'

yyscamper commented on lib/jobs/pdu-node-relations.js:
```diff
@@ -0,0 +1,142 @@
+        var self = this;
+        return waterline.nodes.findByIdentifier(self.nodeId)
+        .then(function (node) {
+            if (node.type !== 'pdu') {
```
Change to use Constant rather than direct string: https://github.com/RackHD/on-core/blob/master/lib/common/constants.js#L316

yyscamper commented on lib/jobs/pdu-node-relations.js:
```diff
@@ -0,0 +1,142 @@
+     * returns a booleen
+     */
+    function _validateIpAddressFormat(ipString) {
+        var regex = /^(?:[0-9]{1,3}\.){3}[0-9]{1,3}$/;
```
There is already a Constant for ipAddress: https://github.com/RackHD/on-core/blob/master/lib/common/constants.js#L215

yyscamper commented on lib/jobs/pdu-node-relations.js:
```diff
@@ -0,0 +1,142 @@
+            })
+            .map(function (result) {
+                //compare hosts
+                var obmServiceHost = _adjustIpStrFormat(result.config.host);
```
Could you please add comment why need do this?

keedya replied:
I do need it to use the filters in promises

yyscamper commented on lib/jobs/pdu-node-relations.js:
```diff
@@ -0,0 +1,142 @@
+    PduRelationsJob.prototype._searchObmServices =
+        function _searchObmServices(pduHost) {
+            var self = this;
+            return Promise.try(function () {
```
You don't need Promise.try as `waterline.obms.find()` has already returned a promise.

keedya replied:
I do need it to use the filters in promises

yyscamper replied:
Ahh, I didn't realize this problem. the promise that returned from waterline isn't powerful than bluebird promise.

yyscamper commented on lib/jobs/pdu-node-relations.js:
```diff
@@ -0,0 +1,142 @@
+                    return waterline.nodes.findByIdentifier(result.node)
+                   .then(function (node) {
+                       if (!node) {
+                           throw new Error('Could not find node with identifier ' + result.node);
```
A little suggestion, if you change previous `findByIdentifier` to `needByIdentifier`, you don't need this throw.

keedya commented:
@yyscamper thanks for your review. I addressed your comments. I will use needByIdentifier in the future. In this task, I wrote unit test around findByIdentifier. I will prefer to leave it in this task unless you think otherwise?
Thanks :)

yyscamper commented on lib/jobs/pdu-node-relations.js:
```diff
@@ -0,0 +1,142 @@
+                   })
+                   .then(function (relations) {
+                       relations.push({
+                           relationType: RELATION_TYPE,
```
If you add `poweredBy` relation to node, then the counterpart `powers` relation should be added into the pdu node. But you forgot the latter relation.

yyscamper replied:
You can reference how the enclosure relation is handled: https://github.com/RackHD/on-tasks/blob/master/lib/jobs/generate-enclosure.js

yyscamper replied:
I think you are copying the design of switch-node-relation job, but it still doesn't resolve the bi-direction relations, so it still need some fix.

The relation is always bi-direction, if you relates A with B by "poweredBy" and update the A's document, then you should still relates B with A by "powers" and update B's document. 

I admit this is a little complicated, but this is exactly the design with enclosure relation. Either build or remove relation is not easy.

I just come up an idea, copy the current `obm` design, detach the relations to a new kind of resource, this will simplify a lot code logic to either build or remove relations.

keedya replied:
@yyscamper I am working on the changes. I will update the PR shortly
Thanks :)

anhou commented:
@keedya  I suggest replacing  findByIdentifier to needByIdentifier in this PR, it's easy to just  replace name findByIdentifier to needByIdentifer in batches in file, and unittest cases wouldn't be affected because the mock method is the same.

anhou commented:
:+1: with comments updated

yyscamper commented:
@keedya : either needByIdentifier or findByIdentifier is OK for me, in your case, you can save some code to use needByIdentfier. If you choose findByIdentifer, don't forget the empty check.

keedya commented:
Address @yyscamper and @anhou comments

keedya commented:
Adding @benbp

yyscamper commented on lib/jobs/pdu-node-relations.js:
```diff
@@ -0,0 +1,191 @@
+
+    PduRelationsJob.prototype.updatePduNodeRelations =
+        function _updatePduNodeRelations(pduNode, poweredNodeId) {
+            Promise.try(function () {
```
This promise need be return

yyscamper replied:
Just like you do in line76

yyscamper commented on lib/jobs/pdu-node-relations.js:
```diff
@@ -0,0 +1,191 @@
+                if (targets.indexOf(poweredNodeId) < 0) {
+                    targets.push(poweredNodeId);
+                }
+                pduRelations.push({
```
If the pduRelations is empty, you can push a new `powers` relation record. But if the pudRelations is not empty, that means you don't need to create a new relation record, just update the existing relation record and insert the new nodeId to targets is enough.

yyscamper commented on lib/jobs/pdu-node-relations.js:
```diff
@@ -0,0 +1,191 @@
+                    relationType: PDU_RELATION_TYPE,
+                    targets: targets
+                });
+                return waterline.nodes.updateByIdentifier(
```
If the relation has already been created and no change, we should prevent from updating the node document. But I see your code logic is whenever the relation is existing or not, create a new one and always update the node document.

yyscamper replied:
The same for function "updatePoweredNodeRelations "

yyscamper commented:
@keedya : Please make sure at least following test cases have been passed using your code:
- [ ] 1 PDU + 1 NODE: no relations before, then after job running, both nodes have been created correct relation.
- [ ]  1 PDU + 2 NODEs: no relations before, then after job running, 3 nodes have been created correct relations. The PDU should have only one power relation with the `targets` contains both nodes.
- [ ] 1 PDU + 1 NODE:  The power relation has existed, after job running, no change for both nodes' document.
- [ ] 1 PDU + 1 NODE:   The PDU has been created relation with another node before, after job running, the PDU node relation will be updated, but shouldn't create another new power relation.

yyscamper commented:
@keedya : Whether you only want to run this job in pdu discovery? If that's true, we don't need to consider too much cases because no power relations existing before your job runs.
If so, please ignore my above 2 comments, the only comment you need fix is let the promise be returned.

:+1: after fix the promise not being returned.

keedya commented:
@yuscamper the logic is to use the create a new one and add existing targets. Same logic. I also return a promise within promise.try()

yyscamper commented:
@keedya : The promise that need to be returned is the promise.try itself, not the one within promise.try. Below function's return value is `undefined` with a side effect of a floating promise, the promise will be run in background.
```javascript
function() {
   Promise.try(...);
}
```

keedya commented:
@yyscamper all set!!

yyscamper commented:
Thanks, :+1:
